### PR TITLE
Also consider call and yield as MIR SSA.

### PR DESCRIPTION
**Pull request by cjgillot**
The SSA analysis on MIR only considered `Assign` statements as defining a SSA local.
This PR adds assignments as part of a `Call` or `Yield` terminator in that category.

This mainly allows to perform CopyProp on a call return place.

The only subtlety is in the dominance property: the assignment is only complete at the beginning of the target block.

### CODE REVIEW

**Comment by fee1-dead** [comment]
Not familiar with this.

r? compiler

**Review by JakobDegen — APPROVED** [review]

**Comment by cjgillot** [comment]
r? compiler

File: compiler/rustc_mir_transform/src/ssa.rs [hunk 1]
```diff
@@ -228,34 +249,34 @@ impl<'tcx> Visitor<'tcx> for SsaVisitor<'_> {
     }
 
     fn visit_place(&mut self, place: &Place<'tcx>, ctxt: PlaceContext, loc: Location) {
-        if place.projection.first() == Some(&PlaceElem::Deref) {
-            // Do not do anything for storage statements and debuginfo.
+        let location = match ctxt {
+            PlaceContext::MutatingUse(
+                MutatingUseContext::Store | MutatingUseContext::Call | MutatingUseContext::Yield,
+            ) => Some(DefLocation::Body(loc)),
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/ssa.rs:255]: @tmiasko Following the discussion on https://github.com/rust-lang/rust/pull/116454#discussion_r1348002248, I reuse `DefLocation::Body` for the terminators. Do you agree with this conclusion?

Comment (cjgillot) [reply compiler/rustc_mir_transform/src/ssa.rs:255]: So let's go small steps. This PR as-is, then https://github.com/rust-lang/rust/commit/abd4ad1b469d827768d556c305c3d7412ef018f9 in a second PR.

**Comment by nnethercote** [comment]
icounts/cycles/wall-time/max-rss are a wash, but binary size and artifact size see some small but clear improvements.

@rustbot label: +perf-regression-triaged

**Pull request merged**